File: DSA/src/Questions/DP/LongestIncreasingSubsequence.java
package Questions.DP;

import java.util.Arrays;

public class LongestIncreasingSubsequence {
    // Problem:- The longest increasing subsequence is described as a subsequence of
    // an array where:

    static int getAns(int[] arr, int n, int ind, int prev_index, int[][] dp) {
        // Base condition
        if (ind == n) {
            return 0;
        }

        if (dp[ind][prev_index + 1] != -1) {
            return dp[ind][prev_index + 1];
        }

        int notTake = getAns(arr, n, ind + 1, prev_index, dp);

        int take = 0;

        if (prev_index == -1 || arr[ind] > arr[prev_index]) {
            take = 1 + getAns(arr, n, ind + 1, ind, dp);
        }

        dp[ind][prev_index + 1] = Math.max(notTake, take);

        return dp[ind][prev_index + 1];
    }

    // Function to find the length of the longest increasing subsequence
    static int lisRecursive(int[] arr, int n) {
        int[][] dp = new int[n][n + 1];

        // Initialize dp array with -1 to mark states as not calculated yet
        for (int[] row : dp) {
            Arrays.fill(row, -1);
        }

        return getAns(arr, n, 0, -1, dp);
    }
    // Time Complexity: O(N*N)

    // Reason: There are N*N states therefore at max ‘N*N’ new problems will be
    // solved.

    // Space Complexity: O(N*N) + O(N)

    // Reason: We are using an auxiliary recursion stack space(O(N)) (see the
    // recursive tree, in the worst case we will go till N calls at a time) and a 2D
    // array ( O(N*N+1))

    static int longestIncreasingSubsequenceTabulation(int[] arr, int n) {

        int[][] dp = new int[n + 1][n + 1];

        for (int ind = n - 1; ind >= 0; ind--) {
            for (int prev_index = ind - 1; prev_index >= -1; prev_index--) {

                int notTake = dp[ind + 1][prev_index + 1];

                int take = 0;

                if (prev_index == -1 || arr[ind] > arr[prev_index]) {

                    take = 1 + dp[ind + 1][ind + 1];
                }

                dp[ind][prev_index + 1] = Math.max(notTake, take);

            }
        }

        return dp[0][0];
    }
    // Time Complexity: O(N*N)

    // Reason: There are two nested loops

    // Space Complexity: O(N*N)

    // Reason: We are using an external array of size ‘(N+1)*(N+1)’. Stack Space is
    // eliminated.

    static int longestIncreasingSubsequenceOptimized(int arr[], int n) {

        int[] next = new int[n + 1];
        int[] cur = new int[n + 1];

        for (int ind = n - 1; ind >= 0; ind--) {
            for (int prev_index = ind - 1; prev_index >= -1; prev_index--) {

                int notTake = 0 + next[prev_index + 1];

                int take = 0;

                if (prev_index == -1 || arr[ind] > arr[prev_index]) {

                    take = 1 + next[ind + 1];
                }

                cur[prev_index + 1] = Math.max(notTake, take);
            }
            next = cur.clone();
        }

        return cur[0];
    }

    // Time Complexity: O(N*N)

    // Reason: There are two nested loops.

    // Space Complexity: O(N)

    // Reason: We are only using two rows of size n.

    public static void main(String[] args) {
        int[] arr = { 10, 22, 9, 33, 21, 50, 41, 60, 80 };
        int n = arr.length;
        System.out.println(lisRecursive(arr, n)); // 6
        System.out.println(longestIncreasingSubsequenceTabulation(arr, n)); // 6
        System.out.println(longestIncreasingSubsequenceOptimized(arr, n)); // 6
    }

}

File: DSA/Python/Questions/DP/longest_increasing_subsequence.py
"""
# Question: Longest Increasing Subsequence
# Link: https://leetcode.com/problems/longest-increasing-subsequence/

# Problem Statement:
# Given an integer array nums, return the length of the longest strictly increasing subsequence.
# A subsequence is a sequence that can be derived from an array by deleting some or no elements
# without changing the order of the remaining elements.

# Example:
# Input: nums = [10,22,9,33,21,50,41,60,80]
# Output: 6
# Explanation: The longest increasing subsequence is [10,22,33,50,60,80]
"""

from typing import List

# 1. Recursive Solution with Memoization
"""
Algorithm:
1. For each element, two choices:
   - Include if greater than previous element
   - Skip current element
2. Track previous index for comparison
3. Use memoization to store results

Time Complexity: O(n^2)
Space Complexity: O(n^2) + O(n) recursion stack
"""


def lis_recursive(arr: List[int]) -> int:
    n = len(arr)
    # Initialize dp array for memoization
    dp = [[-1] * (n + 1) for _ in range(n)]

    def solve(ind: int, prev_ind: int) -> int:
        # Base case: reached end of array
        if ind == n:
            return 0

        # Return memoized result if available
        if dp[ind][prev_ind + 1] != -1:
            return dp[ind][prev_ind + 1]

        # Don't take current element
        length = solve(ind + 1, prev_ind)

        # Take current element if valid
        if prev_ind == -1 or arr[ind] > arr[prev_ind]:
            length = max(length, 1 + solve(ind + 1, ind))

        # Store and return result
        dp[ind][prev_ind + 1] = length
        return length

    return solve(0, -1)


# 2. Tabulation Solution
"""
Algorithm:
1. Create 2D dp array
2. Process array from end to start
3. For each position, try all possible previous elements
4. Track maximum length at each step

Time Complexity: O(n^2)
Space Complexity: O(n^2)
"""


def lis_tabulation(arr: List[int]) -> int:
    n = len(arr)
    # Create dp array
    dp = [[0] * (n + 1) for _ in range(n + 1)]

    # Fill dp table
    for ind in range(n - 1, -1, -1):
        for prev_ind in range(ind - 1, -2, -1):
            # Don't take current element
            length = dp[ind + 1][prev_ind + 1]

            # Take current element if valid
            if prev_ind == -1 or arr[ind] > arr[prev_ind]:
                length = max(length, 1 + dp[ind + 1][ind + 1])

            dp[ind][prev_ind + 1] = length

    return dp[0][0]


# 3. Space Optimized Solution
"""
Algorithm:
1. Use two arrays instead of 2D array
2. Next and current arrays sufficient
3. Update arrays in alternating fashion

Time Complexity: O(n^2)
Space Complexity: O(n)
"""


def lis_optimized(arr: List[int]) -> int:
    n = len(arr)
    # Initialize arrays
    next = [0] * (n + 1)
    curr = [0] * (n + 1)

    # Process array
    for ind in range(n - 1, -1, -1):
        for prev_ind in range(ind - 1, -2, -1):
            # Don't take current element
            length = next[prev_ind + 1]

            # Take current element if valid
            if prev_ind == -1 or arr[ind] > arr[prev_ind]:
                length = max(length, 1 + next[ind + 1])

            curr[prev_ind + 1] = length

        next = curr[:]

    return curr[0]


def main():
    # Test cases
    test_cases = [
        [10, 22, 9, 33, 21, 50, 41, 60, 80],  # Regular case
        [3, 10, 2, 1, 20],  # Multiple increasing sequences
        [7, 7, 7, 7],  # Equal elements
        [1],  # Single element
        [5, 4, 3, 2, 1],  # Decreasing sequence
    ]

    for i, arr in enumerate(test_cases):
        print(f"\nTest Case {i + 1}: {arr}")
        print(f"Recursive Solution: {lis_recursive(arr)}")
        print(f"Tabulation Solution: {lis_tabulation(arr)}")
        print(f"Optimized Solution: {lis_optimized(arr)}")


if __name__ == "__main__":
    main()
